tests/e2e/test_regex_filtering_against_sample_names.py

```diff
@@ -0,0 +1,298 @@
+"""
+End to end tests for monitoring runs to upload with regex patterns
+defined in the config file to limit what runs are uploaded
+
+We will create 3 runs in the same directory with different samplesheets,
+where the sample names in each inform what assay the sequencing is for.
+We will simulate monitoring with 2 patterns to upload from the same
+directory but to upload to 2 different remote paths for run_1 and run_2,
+with run_3 not matching either pattern and therefore should be omitted.
+"""
+
+from argparse import Namespace
+from copy import deepcopy
+from datetime import datetime
+from glob import glob
+import json
+from pathlib import Path
+import unittest
+from unittest.mock import patch
+import os
+import shutil
+
+import boto3
+
+from e2e import BASE_CONFIG, S3_BUCKET, TEST_DATA_DIR
+from e2e.helper import create_files
+from s3_upload.s3_upload import main as s3_upload_main
+
+
+class TestConfigRegexPatternsAgainstSampleNames(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        # create test sequencing runs in one sequencer output directory
+        cls.run_1 = os.path.join(TEST_DATA_DIR, "sequencer_a", "run_1")
+        cls.run_2 = os.path.join(TEST_DATA_DIR, "sequencer_a", "run_2")
+        cls.run_3 = os.path.join(TEST_DATA_DIR, "sequencer_a", "run_3")
+
+        codes = ["assay_1", "assay_2", "assay_3"]
+
+        with open(os.path.join(TEST_DATA_DIR, "example_samplesheet.csv")) as f:
+            base_samplesheet = f.read().splitlines()
+
+        # create as a complete runs with some example files
+        for run_dir, code in zip((cls.run_1, cls.run_2, cls.run_3), codes):
+            create_files(
+                run_dir,
+                "RunInfo.xml",
+                "CopyComplete.txt",
+                "Config/Options.cfg",
+                "InterOp/EventMetricsOut.bin",
+            )
+
+            # example samplesheet has `assay_1` for all samples, replace
+            # this with the current assay code we want to write in
+            run_samplesheet = deepcopy(base_samplesheet)
+            run_samplesheet = [
+                x.replace("assay_1", code) for x in run_samplesheet
+            ]
+
+            with open(os.path.join(run_dir, "samplesheet.csv"), "w") as fh:
+                fh.write("\n".join(run_samplesheet))
+
+        # define separate full unique paths to upload test runs to by
+        # the assay type they're for
+        now = datetime.now().strftime("%y%m%d_%H%M%S")
+        cls.parent_remote_path = f"s3_upload_e2e_test/{now}"
+        cls.run_1_remote_path = f"{cls.parent_remote_path}/assay_1"
+        cls.run_2_remote_path = f"{cls.parent_remote_path}/assay_2"
```

Comment:
Do we not include run_3 here because it won't match sample regex? If it was created, would it remain empty because s3_upload_main would not upload the samples that don't match regex?

Reply:
wait, nevermind! The path wouldn't be created in the first place because it doesn't match regex, so it won't show up in to_upload